<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.489">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>WGS_Bacteria - Example #1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">WGS_Bacteria</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./example_1.html" aria-current="page"> 
<span class="menu-text">Example #1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./example_2.html"> 
<span class="menu-text">Example #2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./setup.html"> 
<span class="menu-text">Set Up</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#conda-env-data-download-qc-reports" id="toc-conda-env-data-download-qc-reports" class="nav-link active" data-scroll-target="#conda-env-data-download-qc-reports">Conda env, data download, QC reports</a></li>
  <li><a href="#trimming-and-post-trim-qc-reports" id="toc-trimming-and-post-trim-qc-reports" class="nav-link" data-scroll-target="#trimming-and-post-trim-qc-reports">Trimming and post trim QC reports</a></li>
  <li><a href="#assembly" id="toc-assembly" class="nav-link" data-scroll-target="#assembly">Assembly</a></li>
  <li><a href="#polishing" id="toc-polishing" class="nav-link" data-scroll-target="#polishing">Polishing</a></li>
  <li><a href="#qc-of-assembly" id="toc-qc-of-assembly" class="nav-link" data-scroll-target="#qc-of-assembly">QC of Assembly</a>
  <ul class="collapse">
  <li><a href="#quality-assessment-tool-for-genome-assemblies-quast" id="toc-quality-assessment-tool-for-genome-assemblies-quast" class="nav-link" data-scroll-target="#quality-assessment-tool-for-genome-assemblies-quast">Quality Assessment Tool for Genome Assemblies (QUAST)</a></li>
  <li><a href="#icarus-quast-contig-browser" id="toc-icarus-quast-contig-browser" class="nav-link" data-scroll-target="#icarus-quast-contig-browser">Icarus QUAST Contig Browser</a></li>
  </ul></li>
  <li><a href="#reorder-contigs" id="toc-reorder-contigs" class="nav-link" data-scroll-target="#reorder-contigs">Reorder contigs</a></li>
  <li><a href="#multi-locus-sequence-typing-mlst-and-antibiotic-resistance" id="toc-multi-locus-sequence-typing-mlst-and-antibiotic-resistance" class="nav-link" data-scroll-target="#multi-locus-sequence-typing-mlst-and-antibiotic-resistance">Multi-Locus Sequence Typing (MLST) and Antibiotic Resistance</a></li>
  <li><a href="#rapid-prokaryotic-genome-annotation" id="toc-rapid-prokaryotic-genome-annotation" class="nav-link" data-scroll-target="#rapid-prokaryotic-genome-annotation">Rapid prokaryotic genome annotation</a></li>
  <li><a href="#comparative-analysis" id="toc-comparative-analysis" class="nav-link" data-scroll-target="#comparative-analysis">Comparative Analysis</a>
  <ul class="collapse">
  <li><a href="#average-nucleotide-identity" id="toc-average-nucleotide-identity" class="nav-link" data-scroll-target="#average-nucleotide-identity">Average Nucleotide Identity</a></li>
  <li><a href="#genome-ring-structures-using-brig" id="toc-genome-ring-structures-using-brig" class="nav-link" data-scroll-target="#genome-ring-structures-using-brig">Genome ring structures using BRIG</a></li>
  <li><a href="#generate-genome-feature-files" id="toc-generate-genome-feature-files" class="nav-link" data-scroll-target="#generate-genome-feature-files">Generate Genome Feature Files</a></li>
  <li><a href="#generate-pangenome" id="toc-generate-pangenome" class="nav-link" data-scroll-target="#generate-pangenome">Generate Pangenome</a></li>
  </ul></li>
  <li><a href="#bug-note" id="toc-bug-note" class="nav-link" data-scroll-target="#bug-note">Bug note</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Example #1</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In short, a complete workflow from Illumina reads to genome assembly, through annotation and comparative genomics.</p>
<p>This example uses just the Illumina data (<a href="https://www.ebi.ac.uk/ena/browser/view/ERR3335404" class="uri">https://www.ebi.ac.uk/ena/browser/view/ERR3335404</a>) generated for the study “<em>Whole genome sequence of Mycobacterium ulcerans CSURP7741, a French Guyana clinical isolate</em>” (<a href="https://www.ebi.ac.uk/ena/browser/view/PRJEB30628" class="uri">https://www.ebi.ac.uk/ena/browser/view/PRJEB30628</a>), “P7741” as a short identity code. <em>Mycobacterium ulcerans</em> is an environmental non-tuberculous mycobacterium responsible for Buruli ulcer.</p>
<p>One reference genome paper is “<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3554023/">Complete genome sequence of the frog pathogen Mycobacterium ulcerans ecovar Liflandii</a>” and it’s sequence <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_020133" class="uri">https://www.ncbi.nlm.nih.gov/nuccore/NC_020133</a>. <em>Mycobacterium ulcerans liflandii 128FXT</em>, complete sequence has NCBI Reference Sequence: NC_020133.1 and is refered to as just “Liflandii” (sequence stored as ./genomes/Liflandii.fasta). Other reference genomes of <em>Mycobacterium ulcerans</em> are utilized in the workflow (Agy99, SGL03 and Shinsuense) and one of <em>Mycobacterium tuberculosis</em> (H37Rv). This work is based upon the excellent presentations and analysis demonstrated by Vincent Appiah (vincentappiah, <a href="https://www.youtube.com/watch?v=S_sRo_85jhs&amp;list=PLe1-kjuYBZ06-7O8eeqLoFsGQYGFbytzd">link</a>) and Daren Ginete (<a href="https://www.youtube.com/watch?v=CTugtM1LR0M">link</a>). The work has been reproduced, adapted and extended in order that it works on a Mac laptop with updated versions of analysis software available in February 2024 and using the Quarto package for presentation and re-usability.</p>
<section id="conda-env-data-download-qc-reports" class="level2">
<h2 class="anchored" data-anchor-id="conda-env-data-download-qc-reports">Conda env, data download, QC reports</h2>
<p>download data and QC of raw Illumina with <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" title="fastqc">fastqc</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use clean install of conda on machine in private space </span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> <span class="at">-V</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># all those envs explained on Set Up page</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env list</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># uncomment appropriately to re-run</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># download data to data </span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># chmod +x download_data.sh</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ./download_data.sh</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># run fastqc</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># mkdir QC_RAW_READS</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co"># fastqc data/*.fastq.gz -o QC_RAW_READS</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># show the raw data</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> data/ <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-nae</span> <span class="st">'print "@F[4..8]\n"'</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>conda 24.1.2
# conda environments:
#
base                  *  /Users/sw/anaconda3
bacterial-genomics-tutorial-sw1     /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw1
bacterial-genomics-tutorial-sw2     /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw2
bacterial-genomics-tutorial-sw3     /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw3
bacterial-genomics-tutorial-sw4     /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw4
bacterial-genomics-tutorial-sw5     /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw5
bacterial-genomics-tutorial-sw7     /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw7
example                  /Users/sw/anaconda3/envs/example
for_drep                 /Users/sw/anaconda3/envs/for_drep
just_quast               /Users/sw/anaconda3/envs/just_quast

    
40M Feb 23 16:02 P7741_R1.fastq.gz
45M Feb 23 16:02 P7741_R2.fastq.gz</code></pre>
</div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_eg1.png" class="img-fluid figure-img" width="800"></p>
<figcaption>Click on links to get the full reports</figcaption>
</figure>
</div>
<p>Results files are QC html reports.</p>
<p><a href="./QC_RAW_READS/P7741_R1_fastqc.html">./QC_RAW_READS/P7741_R1_fastqc.html</a></p>
<p><a href="./QC_RAW_READS/P7741_R1_fastqc.html">./QC_RAW_READS/P7741_R2_fastqc.html</a></p>
</section>
<section id="trimming-and-post-trim-qc-reports" class="level2">
<h2 class="anchored" data-anchor-id="trimming-and-post-trim-qc-reports">Trimming and post trim QC reports</h2>
<p>Using <a href="http://www.usadellab.org/cms/index.php?page=trimmomatic">Trimmomatic: A flexible read trimming tool for Illumina NGS data</a> from Björn Usadel lab (hallöchen!). Alternative is <a href="https://github.com/najoshi/sickle">Sickle</a> (A windowed adaptive trimming tool for FASTQ files using quality)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use clean install of conda on machine in private space </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># uncomment appropriately to re-run</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># mkdir trimmed_reads</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="va">read1</span><span class="op">=</span>data/P7741_R1.fastq.gz</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="va">read2</span><span class="op">=</span>data/P7741_R2.fastq.gz</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="va">OutputForwardPaired</span><span class="op">=</span>trimmed_reads/P7741_R1.fastq.gz    </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="va">OutputForwardUnpaired</span><span class="op">=</span>trimmed_reads/P7741_unpaired_R1.fastq.gz</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="va">OutputReversePaired</span><span class="op">=</span>trimmed_reads/P7741_R2.fastq.gz</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="va">OutputReverseUnpaired</span><span class="op">=</span>trimmed_reads/P7741_unpaired_R2.fastq.gz</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="va">threads</span><span class="op">=</span>4</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="co"># trimmomatic PE -threads $threads $read1 $read2 \</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co"># $OutputForwardPaired $OutputForwardUnpaired \</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co"># $OutputReversePaired $OutputReverseUnpaired \</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:8:true HEADCROP:3 TRAILING:10 MINLEN:25</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="co"># mkdir QC_TRIMMED_READS</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># fastqc trimmed_reads/* -o QC_TRIMMED_READS</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-1</span> QC_TRIMMED_READS</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>P7741_R1_fastqc.html
P7741_R1_fastqc.zip
P7741_R2_fastqc.html
P7741_R2_fastqc.zip
P7741_unpaired_R1_fastqc.html
P7741_unpaired_R1_fastqc.zip
P7741_unpaired_R2_fastqc.html
P7741_unpaired_R2_fastqc.zip</code></pre>
</div>
</div>
<p>Results files are QC html reports. vincentappiah’s results are different around 16:29. In video he speaks of sickle trimming, but the code uses trimmomatic.</p>
<p><a href="./QC_TRIMMED_READS/P7741_R1_fastqc.html">./QC_TRIMMED_READS/P7741_R1_fastqc.html</a></p>
<p><a href="./QC_TRIMMED_READS/P7741_R1_fastqc.html">./QC_TRIMMED_READS/P7741_R2_fastqc.html</a></p>
<p><a href="./QC_TRIMMED_READS/P7741_unpaired_R1_fastqc.html">./QC_TRIMMED_READS/P7741_unpaired_R1_fastqc.html</a></p>
<p><a href="./QC_TRIMMED_READS/P7741_unpaired_R2_fastqc.html">./QC_TRIMMED_READS/P7741_unpaired_R2_fastqc.html</a></p>
</section>
<section id="assembly" class="level2">
<h2 class="anchored" data-anchor-id="assembly">Assembly</h2>
<p>using <a href="https://github.com/ablab/spades?tab=readme-ov-file">SPAdes - St.&nbsp;Petersburg genome assembler (SPAdes)</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Real problems setting up spades with conda - notes on Set Up page</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Solution was to get the spades Mac binaries</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># It runs but takes quite some time on my mini-server</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Therefore delete directory</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co"># P7741_SPADES_OUT</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># and re-run if needed</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ~/SPAdes-3.15.5-Darwin/bin/spades.py --careful -o P7741_SPADES_OUT -1 trimmed_reads/P7741_R1.fastq.gz -2 trimmed_reads/P7741_R2.fastq.gz</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"number of contigs"</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'&gt;'</span> P7741_SPADES_OUT/contigs.fasta <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"number of scaffolds"</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'&gt;'</span> P7741_SPADES_OUT/scaffolds.fasta <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"directory listing"</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-1</span> P7741_SPADES_OUT</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>number of contigs
    3243

number of scaffolds
    3232

directory listing
K21
K33
K55
K77
assembly_graph.fastg
assembly_graph_after_simplification.gfa
assembly_graph_with_scaffolds.gfa
before_rr.fasta
contigs.fasta
contigs.paths
corrected
dataset.info
input_dataset.yaml
misc
mismatch_corrector
params.txt
pipeline_state
run_spades.sh
run_spades.yaml
scaffolds.fasta
scaffolds.paths
spades.log
tmp
warnings.log</code></pre>
</div>
</div>
</section>
<section id="polishing" class="level2">
<h2 class="anchored" data-anchor-id="polishing">Polishing</h2>
<p>with</p>
<ul>
<li><p><a href="https://bio-bwa.sourceforge.net/">bwa</a> (alignment via Burrows-Wheeler transformation)</p></li>
<li><p><a href="https://www.htslib.org/">samtools</a> (for manipulating high-throughput sequencing data)</p></li>
<li><p><a href="https://github.com/broadinstitute/pilon/wiki">pilon</a> (automatically improve draft assemblies and find variation among strains including large event detection)</p></li>
</ul>
<p>Use script to do 4 rounds (not guaranteed to be optimal) of polishing. Runs but takes quite some time on my mini-server but it goes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># To clean re-run delete dir polishing_process</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># re-run script</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ./polish.sh</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co"># list out the files developed </span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="fu">find</span> polishing_process</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>polishing_process
polishing_process/pilon_stage4.changes
polishing_process/pilon_stage1.fasta.ann
polishing_process/pilon_stage3.fasta.amb
polishing_process/pilon_stage3.fasta.bwt
polishing_process/stage1.pilon
polishing_process/pilon_stage1.fasta
polishing_process/pilon_stage3.fasta
polishing_process/raw_assembly.fasta.ann
polishing_process/mapping3.sorted.bam
polishing_process/stage3.pilon
polishing_process/pilon_stage4.fasta
polishing_process/mapping4.sorted.bam
polishing_process/pilon_stage2.fasta.amb
polishing_process/stage4.pilon
polishing_process/raw_assembly.fasta
polishing_process/pilon_stage2.fasta.bwt
polishing_process/raw_assembly.fasta.pac
polishing_process/raw_assembly.fasta.sa
polishing_process/pilon_stage2.fasta
polishing_process/pilon_stage1.fasta.pac
polishing_process/stage2.pilon
polishing_process/mapping1.sorted.bam
polishing_process/pilon_stage1.fasta.sa
polishing_process/pilon_stage3.fasta.ann
polishing_process/pilon_stage1.fasta.amb
polishing_process/mapping1.sorted.bam.bai
polishing_process/mapping2.sorted.bam
polishing_process/pilon_stage1.fasta.bwt
polishing_process/pilon_stage3.changes
polishing_process/raw_assembly.fasta.bwt
polishing_process/pilon_stage2.fasta.pac
polishing_process/pilon_stage2.changes
polishing_process/raw_assembly.fasta.amb
polishing_process/pilon_stage2.fasta.ann
polishing_process/pilon_stage3.fasta.sa
polishing_process/pilon_stage2.fasta.sa
polishing_process/mapping2.sorted.bam.bai
polishing_process/pilon_stage3.fasta.pac
polishing_process/mapping3.sorted.bam.bai
polishing_process/mapping4.sorted.bam.bai
polishing_process/pilon_stage1.changes</code></pre>
</div>
</div>
</section>
<section id="qc-of-assembly" class="level2">
<h2 class="anchored" data-anchor-id="qc-of-assembly">QC of Assembly</h2>
<p>using <a href="https://quast.sourceforge.net/">quast</a> with results linked in below</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># env just with the quast software</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate just_quast</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co"># genomes dir taken from vincentappiah repo</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># quast throws an error with default script</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co"># -t 1 solves </span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># other potential issue may exist as the install of quast was not completed...</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co"># see SetUp page</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># uncomment to run again</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co"># mkdir QC_ASSEMBLY</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="co"># quast.py -t 1 -o QC_ASSEMBLY -R genomes/Liflandii.fasta P7741_SPADES_OUT/scaffolds.fasta P7741.polished.fasta</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Results notes</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="co"># N50 quality didn't change with polishing, GC% didn't change,</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Slight change in misassemblies </span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a><span class="co"># list out the top dir </span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-1</span> QC_ASSEMBLY</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>aligned_stats
basic_stats
contigs_reports
genome_stats
icarus.html
icarus_viewers
quast.log
report.html
report.pdf
report.tex
report.tsv
report.txt
transposed_report.tex
transposed_report.tsv
transposed_report.txt</code></pre>
</div>
</div>
<section id="quality-assessment-tool-for-genome-assemblies-quast" class="level3">
<h3 class="anchored" data-anchor-id="quality-assessment-tool-for-genome-assemblies-quast">Quality Assessment Tool for Genome Assemblies (QUAST)</h3>
<p><a href="QC_ASSEMBLY/report.html"><img src="img/quast_eg1.png" class="img-fluid" width="800" alt="QUAST view. Click on link to get interactive html report."></a></p>
<p><a href="QC_ASSEMBLY/report.html">QC_ASSEMBLY/report.html</a></p>
</section>
<section id="icarus-quast-contig-browser" class="level3">
<h3 class="anchored" data-anchor-id="icarus-quast-contig-browser">Icarus QUAST Contig Browser</h3>
<p><a href="QC_ASSEMBLY/icarus.html" title="QC_ASSEMBLY/icarus.html">QC_ASSEMBLY/icarus.html</a></p>
<p><a href="QC_ASSEMBLY/icarus_viewers/contig_size_viewer.html">QC_ASSEMBLY/icarus_viewers/contig_size_viewer.html</a></p>
<p><a href="QC_ASSEMBLY/icarus_viewers/contig_size_viewer.html"><img src="img/icarus_eg1.png" class="img-fluid" width="800" alt="ICARUS view. Click on link or the image to get the interactive reports."></a></p>
<p><a href="QC_ASSEMBLY/icarus_viewers/alignment_viewer.html" title="QC_ASSEMBLY/icarus_viewers/alignment_viewer.html">QC_ASSEMBLY/icarus_viewers/alignment_viewer.html</a></p>
<p><a href="QC_ASSEMBLY/icarus_viewers/alignment_viewer.html"><img src="img/icarus_eg2.png" class="img-fluid" width="800" alt="ICARUS view. Click on link or the image to get the interactive reports."></a></p>
</section>
</section>
<section id="reorder-contigs" class="level2">
<h2 class="anchored" data-anchor-id="reorder-contigs">Reorder contigs</h2>
<p>using ragtag</p>
<p>apart from the paper listed below, there is a more recent publication</p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02823-7">Automated assembly scaffolding using RagTag elevates a new tomato system for high-throughput genome editing</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># genomes/Agy99.fasta is NC_008611.1 Mycobacterium ulcerans Agy99, complete sequence</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># note this is a different genome to genomes/Liflandii.fasta</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="ex">ragtag.py</span> <span class="at">-c</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co"># uncomment to re-run</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co">## ragtag.py scaffold genomes/Agy99.fasta P7741.polished.fasta -o P7741_reordered</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the reordered contig with a custom (modified!) python script</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="co"># The scripts accept name of the ragtag file containing the reordered contigs </span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="co"># and accession number for the reference genome</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> extract_reordered.py P7741_reordered/ragtag.scaffold.fasta NC_008611.1</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a><span class="co"># at this stage vincentappiah gets length of 5291728 </span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a><span class="co"># and a gc of 64.98</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a><span class="co"># potentially the difference is due to newer version of ragtag</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. v1.0.2 vs RagTag v2.1.0</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Alonge, Michael, et al. "RaGOO: fast and accurate reference-guided scaffolding of draft genomes."
Genome biology 20.1 (2019): 1-17.
https://doi.org/10.1186/s13059-019-1829-6
    
GC Percent, opt=ignore: 65.073454
GC Percent, opt=remove: 65.701833
GC Percent, opt=weighted: 65.551660
Sequence Length: 5012494 bp
draft genome sequence extracted</code></pre>
</div>
</div>
</section>
<section id="multi-locus-sequence-typing-mlst-and-antibiotic-resistance" class="level2">
<h2 class="anchored" data-anchor-id="multi-locus-sequence-typing-mlst-and-antibiotic-resistance">Multi-Locus Sequence Typing (MLST) and Antibiotic Resistance</h2>
<p>using <a href="https://github.com/tseemann/mlst" class="uri">https://github.com/tseemann/mlst</a> and <a href="https://github.com/tseemann/abricate" class="uri">https://github.com/tseemann/abricate</a></p>
<p>The results show the expected identity (7 alleles of mycobacteria) and there is an antibiotic resistance gene.</p>
<p>But need to use cgMLST, wgMLST and more…</p>
<p>Need to look at “chewBBACA is a comprehensive pipeline including a set of functions for the creation and validation of whole genome and core genome MultiLocus Sequence Typing (wg/cgMLST) schemas..”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Alleles"</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="ex">mlst</span> <span class="at">--csv</span> P7741.reordered.fasta <span class="op">&gt;</span> mlst.csv</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Antibiotic resistance"</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="ex">abricate</span> P7741.reordered.fasta <span class="op">&gt;</span> amr.summary.tab</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> amr.summary.tab</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Alleles

[11:44:55] This is mlst 2.23.0 running on darwin with Perl 5.030003
[11:44:55] Checking mlst dependencies:
[11:44:55] Found 'blastn' =&gt; /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw7/bin/blastn
[11:44:55] Found 'any2fasta' =&gt; /Users/sw/anaconda3/envs/bacterial-genomics-tutorial-sw7/bin/any2fasta
[11:44:55] Found blastn: 2.15.0+ (002015)
[11:44:55] Excluding 3 schemes: ecoli abaumannii vcholerae_2
[11:44:57] Found exact allele match mycobacteria_2.S7-10
[11:44:57] Found exact allele match mycobacteria_2.L16-450
[11:44:57] Found exact allele match mycobacteria_2.S12-394
[11:44:57] Found exact allele match mycobacteria_2.L19-11
[11:44:57] Found exact allele match mycobacteria_2.S19-13
[11:44:57] Found exact allele match mycobacteria_2.L35-10
[11:44:57] Found exact allele match mycobacteria_2.S14Z-10
[11:44:57] mlst also supports --json output for the modern bioinformatician.
[11:44:57] Done.

Antibiotic resistance

Using nucl database ncbi:  5386 sequences -  2023-Nov-4
Processing: P7741.reordered.fasta
Found 1 genes in P7741.reordered.fasta
Tip: you can use the --summary option to combine reports in a presence/absence matrix.
Done.
#FILE   SEQUENCE    START   END STRAND  GENE    COVERAGE    COVERAGE_MAP    GAPS    %COVERAGE   %IDENTITY   DATABASE    ACCESSION   PRODUCT RESISTANCE
P7741.reordered.fasta   P7741   1110389 1110934 -   aac(2')-Ic  1-546/546   ========/====== 2/2 99.82   82.27   ncbi    NG_047229.1 aminoglycoside N-acetyltransferase AAC(2')-Ic   GENTAMICIN;TOBRAMCYIN</code></pre>
</div>
</div>
</section>
<section id="rapid-prokaryotic-genome-annotation" class="level2">
<h2 class="anchored" data-anchor-id="rapid-prokaryotic-genome-annotation">Rapid prokaryotic genome annotation</h2>
<p>using <a href="https://github.com/tseemann/prokka" class="uri">https://github.com/tseemann/prokka</a> also from Torsten Seemann</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="va">cpus</span><span class="op">=</span>4</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co"># uncomment to re-run</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ran is around 10 minutes with 4 cores</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">#prokka --cpus $cpus --kingdom Bacteria --locustag P7741 --outdir P7741_annotation --prefix P7741 --addgenes P7741.reordered.fasta</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co"># stats</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Summary"</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> P7741_annotation/P7741.txt </span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="co"># get some counts of genome features</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="co"># vincentappiah numbers differ for reason speculated over in above sections</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Another bespoke summary"</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> get_annot_stats.py P7741_annotation P7741</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="co"># show first 3 protein sequences</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"First 3 protein sequences"</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-13</span> P7741_annotation/P7741.faa</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="co"># show some 3 genes/CDS</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"First 3 genes / CDS"</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-7</span> P7741_annotation/P7741.tsv</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="co"># show some pseudogenes</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Some pseudogenes"</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a><span class="ex">./get_pseudo.pl</span> P7741_annotation/P7741.faa <span class="op">&gt;</span> P7741_annotation/P7741.pseudo.txt</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-10</span> P7741_annotation/P7741.pseudo.txt</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"...."</span></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Summary
organism: Genus species strain 
contigs: 1
bases: 5012494
CDS: 4805
gene: 4852
rRNA: 3
tRNA: 43
tmRNA: 1

Another bespoke summary
Counting Annotated Features.............................
CDS:4805
gene:4852
tRNA:43
tmRNA:1
rRNA:3
IS110 family transposase IS117:1
IS30 family transposase ISMsm8:1
Insertion_Sequences:2
Pseudogenes:141

First 3 protein sequences
&gt;P7741_00001 Anti-sigma-M factor RsmA
MHITVSTPPPVIPLSHDQVLDLLQRAPDYGPFADPSRRASCLNGLGYPASTPILGAQPVD
INARPGVLLVLAGDAPADLAVYAVALNCSAADTGLLASTTLPRLPDS
&gt;P7741_00002 Thioredoxin reductase
MTTDSSADATIHDVIVIGSGPAGYTAALYTARAQLAPLVFEGTSFGGALMTTTEVENYPG
FREGITGPELMDEMREQALRFGADLRMEDVQSVSLDGPIKSVVTSDGETHRARAVILAMG
AAARYLHVPGEQELLGRGVSSCATCDGFFFRDQDIAVIGGGDSAMEEATFLTRFARSVTL
VHRRDEFRASKIMINRAQANDKIRILTNKIVLAVDGETTVTGLQLRDTVTGEETTLPVTG
VFVAIGHEPRSSLVRDAVDVDPDGYVLVNGRTTGTSLEGVFAAGDLVDRTYRQAVTAAGS
GCAAAIDAERWLAEHEATGDADSTDTLIGAQQ
&gt;P7741_00003 Thioredoxin
MTDSEKSSATIEVSDASFSTEVLSSNKPVLVDFWATWCGPCKMVAPVLEEIATERADNLT
VAKLDVDANPETARNFQVVSIPTMILFKDGEPVKRIVGAKGKAALLRELSDAVPNLG

First 3 genes / CDS
locus_tag   ftype   length_bp   gene    EC_number   COG product
P7741_00001 gene    324 rsmA_1          
P7741_00001 CDS 324 rsmA_1          Anti-sigma-M factor RsmA
P7741_00002 gene    999 trxB_1          
P7741_00002 CDS 999 trxB_1  1.8.1.9 COG0492 Thioredoxin reductase
P7741_00003 gene    354 trxA            
P7741_00003 CDS 354 trxA        COG0526 Thioredoxin

Some pseudogenes
Found 4805 genes.
Found 141 potential psuedo-genes
P7741_00021 &amp; P7741_00022 =&gt; HTH-type transcriptional activator TipA
P7741_00062 &amp; P7741_00063 =&gt; putative protein
P7741_00063 &amp; P7741_00064 =&gt; putative protein
P7741_00161 &amp; P7741_00162 =&gt; ESX-1 secretion-associated protein EspK
P7741_00272 &amp; P7741_00273 =&gt; putative protein
P7741_00273 &amp; P7741_00274 =&gt; putative protein
P7741_00276 &amp; P7741_00277 =&gt; putative monooxygenase
P7741_00279 &amp; P7741_00280 =&gt; Putative citrate synthase 2
P7741_00281 &amp; P7741_00282 =&gt; Sphingomyelinase
P7741_00293 &amp; P7741_00294 =&gt; putative protein
....</code></pre>
</div>
</div>
</section>
<section id="comparative-analysis" class="level2">
<h2 class="anchored" data-anchor-id="comparative-analysis">Comparative Analysis</h2>
<section id="average-nucleotide-identity" class="level3">
<h3 class="anchored" data-anchor-id="average-nucleotide-identity">Average Nucleotide Identity</h3>
<ul>
<li><p>dRep is a python program for rapidly comparing large numbers of genomes</p></li>
<li><p>Comments in dendrogram.sh says it needs <a href="https://github.com/ParBLiSS/FastANI" class="uri">https://github.com/ParBLiSS/FastANI</a> and that leads to <a href="https://drep.readthedocs.io/en/latest/installation.html" class="uri">https://drep.readthedocs.io/en/latest/installation.html</a></p></li>
<li><p>FastANI is developed for fast alignment-free computation of whole-genome Average Nucleotide Identity (ANI).</p></li>
<li><p>Script dendrogram.sh calls <a href="https://github.com/MrOlm/drep">dRep</a>: it is installed in conda bacterial-genomics-tutorial-sw7 but not possible to add fastani to this build do to errors (with perl versions if I remember correctly), so build a particular environment for dRep (see Set Up page).</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># particular environment for dRep</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate for_drep</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co"># dRep check_dependencies</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co"># uncomment to re-run</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ./dendogram.sh</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>Some of the outputs from the PDFs generated are given as screen-shot images here.</p></li>
<li><p>Cluster of P7741/Liflandii/Shinsuense is different between methods ANI methods but agreement with the distinction of SGL03 and Agy99 that are approaching 100% Average Nucleotide Identity.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/cluster1.png" class="img-fluid figure-img" width="800"></p>
<figcaption>fastANI cluster: just the <em>Mycobacterium ulcerans</em> genomes</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/cluster2.png" class="img-fluid figure-img" width="800"></p>
<figcaption>MASH cluster: H37Rv (green label - hard to read in image) is <em>Mycobacterium tuberculosis</em> and is clearly very different, an outlier with respect to ANI, to the <em>Mycobacterium ulcerans</em> genomes.</figcaption>
</figure>
</div>
</section>
<section id="genome-ring-structures-using-brig" class="level3">
<h3 class="anchored" data-anchor-id="genome-ring-structures-using-brig">Genome ring structures using BRIG</h3>
<p>using <a href="https://github.com/happykhan/BRIG" class="uri">https://github.com/happykhan/BRIG</a></p>
<ul>
<li><p>BRIG is a standalone interactive application with interface to run BLAST and to render the output as ring structure visualizations.</p></li>
<li><p>get BRIG-0.95-dist: start with enough RAM, java -Xmx1500M -jar BRIG.jar</p></li>
<li><p>Needs Genbank formats as generated by prokka (e.g.&nbsp;P7741.gbk).</p>
<ul>
<li><p>This required re-running prokka for reference genomes as get_genome_gffs.sh previously deleted the analysis folder after taking gff</p></li>
<li><p>Now I keep the complete folder for future usage</p></li>
</ul></li>
<li><p>The forms in BRIG need acclimatization, but once understood, the tool works is great</p>
<ul>
<li><p>the path to BLAST binary folder needs to be set in “Configure BRIG options…”</p>
<ul>
<li><p>in this case, pairwise genome blastn jobs are performed</p></li>
<li><p>compute job progress can be seen in the specified “Output folder”</p>
<ul>
<li><p>e.g.&nbsp;WGS_Bacteria/BRIG/Mycobacterium_ulcerans_5g</p></li>
<li><p>folder I used for the 5 genomes of Mycobacterium ulcerans</p></li>
</ul></li>
</ul></li>
<li><p>it’s useful to set-up a symbolic link at same level as BRIG install to gkb directory</p></li>
<li><p>extensive image rendering options are available to customize the graphic</p></li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># find gbks and copy to folder </span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co"># mkdir gbk</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># find . -name "*.gbk" | perl -ne 'chomp ; print "cp $_ gbk/\n"' -</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># could redirect to a shell script, but prefer to develop commands and cut/paste into terminal for short batch processes </span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./H37Rv/H37Rv.gbk gbk/</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./SGL03/SGL03.gbk gbk/</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./Shinsuense/Shinsuense.gbk gbk/</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./P7741_annotation/P7741.gbk gbk/</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./Agy99/Agy99.gbk gbk/</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./Liflandii/Liflandii.gbk gbk/</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Mycobacterium ulcerans: 5 genomes compared - P7741 was assembled in the workflow. Click on image gets the SVG version.</li>
</ul>
<p><a href="img/My_ulcerans_5g.svg"><img src="img/My_ulcerans_5g.png" class="img-fluid" width="800" alt="Mycobacterium ulcerans: 5 genomes compared - P7741 was assembled in the workflow. Click on image gets the SVG version."></a></p>
<ul>
<li>Mycobacterium ulcerans liflandii vs Mycobacterium tuberculosis (var. H37Rv). Click on image gets the SVG version.</li>
</ul>
<p><a href="img/ulcerans_vs_tuberculosis.svg"><img src="img/ulcerans_vs_tuberculosis.png" class="img-fluid" width="800" alt="Mycobacterium ulcerans liflandii vs Mycobacterium tuberculosis. Click on image gets the SVG version."></a></p>
</section>
<section id="generate-genome-feature-files" class="level3">
<h3 class="anchored" data-anchor-id="generate-genome-feature-files">Generate Genome Feature Files</h3>
<p>using <a href="https://github.com/tseemann/prokka" class="uri">https://github.com/tseemann/prokka</a> to run 5 reference genomes Agy99, Liflandii, SGL03, Shinsuense, H37Rv</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># takes just over an hour on laptop to </span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co"># generate gffs for 6 bacterial genomes</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ./get_genome_gffs.sh </span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> gffs <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{print $5 " " $6 $7 " " $8" " $9}'</span> </span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   
7.2M Feb28 18:55 Agy99.gff
5.5M Feb28 19:28 H37Rv.gff
8.0M Feb28 19:04 Liflandii.gff
6.4M Feb28 19:28 P7741.gff
7.2M Feb28 19:12 SGL03.gff
7.6M Feb28 19:21 Shinsuense.gff</code></pre>
</div>
</div>
</section>
<section id="generate-pangenome" class="level3">
<h3 class="anchored" data-anchor-id="generate-pangenome">Generate Pangenome</h3>
<p>using</p>
<ul>
<li><p>roary and roary_plots.py <a href="https://sanger-pathogens.github.io/Roary/" class="uri">https://sanger-pathogens.github.io/Roary/</a></p>
<ul>
<li>Roary is a high speed stand alone pan genome pipeline, which takes annotated assemblies in GFF3 format (produced by Prokka (Seemann, 2014)) and calculates the pan genome.</li>
</ul></li>
<li><p>FastTree <a href="https://microbesonline.org/fasttree/" class="uri">https://microbesonline.org/fasttree/</a></p>
<ul>
<li>FastTree infers approximately-maximum-likelihood phylogenetic trees from alignments of nucleotide or protein sequences.</li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conda </span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/anaconda3/etc/profile.d/conda.sh</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bacterial-genomics-tutorial-sw7</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co"># uncomment to re-run</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ./get_pangenome.sh </span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Pangenome analysis using roary and fasttree</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co"># i have 4 cores</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="va">threads</span><span class="op">=</span>4</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co"># # Uncomment appropriately to re-run</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co"># roary -f pangenome -p $threads -e -n -v --mafft gffs/*.gff</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="co"># # Generate alignment file</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="co"># FastTree -nt -gtr pangenome/core_gene_alignment.aln &gt; pangenome/mytree.newick</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a><span class="co"># # Plot phylogenetic tree and presence/absense in svg format</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a><span class="co"># python roary_plots.py --labels --format svg pangenome/mytree.newick pangenome/gene_presence_absence.csv</span></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="co"># # same as png</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="co"># python roary_plots.py --labels              pangenome/mytree.newick pangenome/gene_presence_absence.csv</span></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="co"># # move all pangenome plots to img folder</span></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="co"># mv pangenome_*.{svg,png} img</span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a><span class="co"># # generates gene_count_summary.png that needs to be renamed and moved </span></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a><span class="co"># # 3 closely related genomes</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a><span class="co"># python gene_count_summary.py P7741 Agy99 Liflandii pangenome/gene_presence_absence.csv</span></span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a><span class="co"># mv gene_count_summary.png  img/P7741_Agy99_Liflandii.png</span></span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a><span class="co"># # 2 closely related genomes and Mtycobacterium tuberculosis</span></span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a><span class="co"># python gene_count_summary.py P7741 H37Rv Liflandii pangenome/gene_presence_absence.csv</span></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a><span class="co"># mv gene_count_summary.png  img/P7741_H37Rv_Liflandii.png</span></span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Genomes used with Roary:"</span></span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-1</span> gffs/<span class="pp">*</span>.gff</span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"pangenome/summary_statistics.txt"</span></span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> pangenome/summary_statistics.txt</span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Genomes used with Roary:
gffs/Agy99.gff
gffs/H37Rv.gff
gffs/Liflandii.gff
gffs/P7741.gff
gffs/SGL03.gff
gffs/Shinsuense.gff

pangenome/summary_statistics.txt
Core genes  (99% &lt;= strains &lt;= 100%)    91
Soft core genes (95% &lt;= strains &lt; 99%)  0
Shell genes (15% &lt;= strains &lt; 95%)  11844
Cloud genes (0% &lt;= strains &lt; 15%)   0
Total genes (0% &lt;= strains &lt;= 100%) 11935</code></pre>
</div>
</div>
<p><a href="pangenome/gene_presence_absence.csv"><img src="img/genes_presence_absense_eg.png" class="img-fluid"></a></p>
<p>Image above: A sample of genes_presence_absense.csv where number of present = 5 genomes. Click on image to get the full CSV file - 11936 lines.</p>
<p><a href="img/pangenome_matrix.svg"><img src="img/pangenome_matrix.png" class="img-fluid" width="5100"></a></p>
<ul>
<li>Image above: Pangenome Tree and Matrix of the Mycobacterium tuberculosis genome versus the Mycobacterium ulcerans genomes. Of the 11935 only 91 are considered core genes across all 6 genomes (a score of between &gt;= 99% similarity). A click on the image gets the scalable SVG version.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pangenome_frequency.png" class="img-fluid figure-img" width="1050"></p>
<figcaption>Pangenome Frequency.</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/P7741_Agy99_Liflandii.png" class="img-fluid figure-img"></p>
<figcaption>Genes unique to the set P7741, Agy99 and Liflandii (1395 = 390 + 2 + 120 + 141 + 3 + 6 + 733). Within the set, for example Agy99 has 120 unique genes. Of the 1395 unique genes within the set, Agy99 and P7741 share only 5, whilst P7741 and Liflandii share 144. Seems like P7741 and Liflandii are phylogenetically closer than to each other than to Agy99.</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/P7741_H37Rv_Liflandii.png" class="img-fluid figure-img"></p>
<figcaption>Genes unique to the set P7741, H37Rv, Liflandii (5169 = 390 + 0 + 3904 + 141 + 1 + 0 + 733). Mycobacterium tuberculosis (var. H37Rv) is clearly, and as expected, relatively distant to the Mycobacterium ulcerans genomes.</figcaption>
</figure>
</div>
</section>
</section>
<section id="bug-note" class="level2">
<h2 class="anchored" data-anchor-id="bug-note">Bug note</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># With Quarto, the strange bug means that if an {r} executable block is included</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># then the preceding {zsh} or {bash} blocks are executed, otherwise not!</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Simply include this and all is executed as desired.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        for (let i = 0; i < 2; i++) {
          container.appendChild(note.children[i].cloneNode(true));
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>