---
title: "Example #1"
---

This example uses just the Illumina data (<https://www.ebi.ac.uk/ena/browser/view/ERR3335404>) generated for the study "**Whole genome sequence of Mycobacterium ulcerans CSURP7741, a French Guyana clinical isolate**" (<https://www.ebi.ac.uk/ena/browser/view/PRJEB30628>). Mycobacterium ulcerans is an environmental non-tuberculous mycobacterium responsible for Buruli ulcer.

The reference genome paper is "[Complete genome sequence of the frog pathogen Mycobacterium ulcerans ecovar Liflandii](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3554023/)" and it's sequence <https://www.ncbi.nlm.nih.gov/nuccore/NC_020133>. Mycobacterium liflandii 128FXT, complete sequence NCBI Reference Sequence: NC_020133.1

This work is based on the excellent presentation and analayis as instructed by vincentappiah [video](https://www.youtube.com/watch?v=S_sRo_85jhs&list=PLe1-kjuYBZ06-7O8eeqLoFsGQYGFbytzd). The work has been reproduced and adapted in order that it works on a Mac laptop with updated versions of analysis software available in 2024 and using the Quarto package for presentation and re-usability.

## Enter Conda env, data download, QC reports

```{zsh}
# use clean install of conda on machine in private space 
source /Users/sw/anaconda3/etc/profile.d/conda.sh
conda -V
conda env list
conda activate bacterial-genomics-tutorial-sw7
# download data to data folder if not done already
chmod +x download_data.sh
# uncomment to re-run
# ./download_data.sh
# run fastqc
mkdir QC_RAW_READS
# fastqc data/*.fastq.gz -o QC_RAW_READS
conda deactivate
```

Results files are QC html reports

[./QC_RAW_READS/P7741_R1_fastqc.html](./QC_RAW_READS/P7741_R1_fastqc.html)

[./QC_RAW_READS/P7741_R2_fastqc.html](./QC_RAW_READS/P7741_R1_fastqc.html)

## Trim and QC reports

```{zsh}
# use clean install of conda on machine in private space 
source /Users/sw/anaconda3/etc/profile.d/conda.sh
conda -V
conda activate bacterial-genomics-tutorial-sw7

mkdir trimmed_reads
read1=data/P7741_R1.fastq.gz
read2=data/P7741_R2.fastq.gz
OutputForwardPaired=trimmed_reads/P7741_R1.fastq.gz    
OutputForwardUnpaired=trimmed_reads/P7741_unpaired_R1.fastq.gz
OutputReversePaired=trimmed_reads/P7741_R2.fastq.gz
OutputReverseUnpaired=trimmed_reads/P7741_unpaired_R2.fastq.gz
threads=4

# uncomment to run
# trimmomatic PE -threads $threads $read1 $read2 \
$OutputForwardPaired $OutputForwardUnpaired \
$OutputReversePaired $OutputReverseUnpaired \
ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:8:true HEADCROP:3 TRAILING:10 MINLEN:25

mkdir QC_TRIMMED_READS
# uncomment to run
# fastqc trimmed_reads/* -o QC_TRIMMED_READS

ls -1 QC_TRIMMED_READS

conda deactivate
```

Results files are QC html reports. Vincent's results are different around 16:29. In video he says sickle trimming was used but the code uses trimmomatic.

[./QC_TRIMMED_READS/P7741_R1_fastqc.html](./QC_TRIMMED_READS/P7741_R1_fastqc.html)

[./QC_TRIMMED_READS/P7741_R2_fastqc.html](./QC_TRIMMED_READS/P7741_R1_fastqc.html)

[./QC_TRIMMED_READS/P7741_unpaired_R1_fastqc.html](./QC_TRIMMED_READS/P7741_unpaired_R1_fastqc.html)

[./QC_TRIMMED_READS/P7741_unpaired_R2_fastqc.html](./QC_TRIMMED_READS/P7741_unpaired_R2_fastqc.html)

## Assembly

```{zsh}
# Real problems setting up spades with conda - notes on Set Up page
# Solution was to get the spades Mac binaries
#
# It runs but takes quite some time on my mini-server
# Therefore delete directory
# P7741_SPADES_OUT
# and re-run if needed
# ~/SPAdes-3.15.5-Darwin/bin/spades.py --careful -o P7741_SPADES_OUT -1 trimmed_reads/P7741_R1.fastq.gz -2 trimmed_reads/P7741_R2.fastq.gz

ls -1 P7741_SPADES_OUT

```

## Polishing

```{zsh}
# conda 
source /Users/sw/anaconda3/etc/profile.d/conda.sh
conda activate bacterial-genomics-tutorial-sw7
# Use standard script to do 4 rounds (not guaranteed to be optimal) of polishing
# Runs but takes quite some time on my mini-server
# To clean re-run, delete dirs,
# polishing_process
# polished_assembly
# re-run script
# ./polish.sh

ls -1 polishing_process
ls -1 polished_assembly

conda deactivate
```

## QC of Assembly

```{zsh}
# conda
source /Users/sw/anaconda3/etc/profile.d/conda.sh
conda -V
# just the quast software
conda activate just_quast

# genomes dir taked from vincents repo
# quast throws an error with default script
# -t 1 solves 
# other potential issue may exist as the install of quast was not completed...
# see SetUp page

# uncomment to run again
# mkdir QC_ASSEMBLY
# quast.py -t 1 -o QC_ASSEMBLY -R genomes/Liflandii.fasta P7741_SPADES_OUT/scaffolds.fasta P7741.polished.fasta

# Results notes
# N50 quality didn't change with polishing, GC% didn't change,
# Slight change in misassemblies 
# Link in the report.html
# This workflow uses just Illumiuna data

ls -1 QC_ASSEMBLY

conda deactivate
```

## Strange bug

```{r}
# With Quarto, the strange bug means that if an {r} executable block is included
# then the preceding {bash} blocks are executed, otherwise not!
# Simply include this and all is executed as desired.
```
